src/Umbraco.Web.Common/Filters/ModelBindingExceptionFilter.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System;
+using System.Net;
+using System.Text.RegularExpressions;
+using Microsoft.AspNetCore.Http.Extensions;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Umbraco.Core;
+using Umbraco.Core.Configuration;
+using Umbraco.Core.Models.PublishedContent;
+using Umbraco.Web.Common.ModelBinders;
+
+namespace Umbraco.Web.Common.Filters
+{
+    /// <summary>
+    ///     An exception filter checking if we get a <see cref="ModelBindingException" /> or <see cref="InvalidCastException" /> with the same model.
+    ///     In which case it returns a redirect to the same page after 1 sec if not in debug mode.
+    /// </summary>
+    /// <remarks>
+    /// This is only enabled when running PureLive
+    /// </remarks>
+    internal class ModelBindingExceptionFilter : ActionFilterAttribute, IExceptionFilter
+    {
+        private static readonly Regex _getPublishedModelsTypesRegex = new Regex("Umbraco.Web.PublishedModels.(\\w+)", RegexOptions.Compiled);
+
+        private readonly IExceptionFilterSettings _exceptionFilterSettings;
+        private readonly IPublishedModelFactory _publishedModelFactory;
+
+        public ModelBindingExceptionFilter(IExceptionFilterSettings exceptionFilterSettings, IPublishedModelFactory publishedModelFactory)
+        {
+            _exceptionFilterSettings = exceptionFilterSettings;
+            _publishedModelFactory = publishedModelFactory ?? throw new ArgumentNullException(nameof(publishedModelFactory));
+        }
+
+        public void OnException(ExceptionContext filterContext)
+        {
+            var disabled = _exceptionFilterSettings?.Disabled ?? false;
+            if (_publishedModelFactory.IsLiveFactory()
+                && !disabled
+                && !filterContext.ExceptionHandled
+                && ((filterContext.Exception is ModelBindingException || filterContext.Exception is InvalidCastException)
+                    && IsMessageAboutTheSameModelType(filterContext.Exception.Message)))
+            {
+                filterContext.HttpContext.Response.Headers.Add(HttpResponseHeader.RetryAfter.ToString(), "1");
+                filterContext.Result = new RedirectResult(filterContext.HttpContext.Request.GetDisplayUrl(), false);
```

Comment:
I think we need to use `GetEncodedUrl()` here:
https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.extensions.urihelper.getdisplayurl?view=aspnetcore-3.1

Reply:
Agreed, have updated.